File: leetcode/349.intersection-of-two-arrays/ts/intersection.ts
const intersection = (nums1: number[], nums2: number[]): number[] => {
    const firstNumbers = new Set<number>()

    for (const n of nums1) {
        firstNumbers.add(n)
    }

    const sharedNumbers = new Set<number>()

    for (const n of nums2) {
        if (firstNumbers.has(n)) sharedNumbers.add(n)
    }

    return Array.from(sharedNumbers)
};

export default intersection

File: leetcode/349.intersection-of-two-arrays/go/intersection.go
package intersectionoftwoarrays

func intersection(nums1 []int, nums2 []int) []int {
	firstSet := make(map[int]struct{})
	intersectionSet := make(map[int]struct{})

	for _, n := range nums1 {
		firstSet[n] = struct{}{}
	}

	for _, n := range nums2 {
		if _, ok := firstSet[n]; ok {
			intersectionSet[n] = struct{}{}
		}
	}

	keys := make([]int, 0, len(intersectionSet))
	for k := range intersectionSet {
		keys = append(keys, k)
	}
	return keys
}
